### ISSUE TITLE
Doom failing to start after adding local package (due to autoloads)

### ISSUE THREAD

**Issue by marcbowes** [hlissner/doom-emacs#4142]
Background: I maintain a private package at my company. One of my users reported problems loading this package under doom. I'm not a doom expert, but have been able to reproduce the problem.

**What did you expect to happen?**
Using the guide at https://github.com/hlissner/doom-emacs/blob/develop/docs/getting_started.org#usingloading-local-packages:
```
 (package! blahblah
 	  :recipe (:local-repo "~/blahblah"))
```

I expect the package to be available after a doom reload or sync + restart.

**What actually happened?**
Doom fails to load. Also, doom sync also fails. The issue is in the autoloads.

If I run Emacs with `--debug-init` and then load each of the .el files for this project immediately after hitting the debugger, everything works just fine.

I'm pretty sure something is wrong with my package but I have no idea what it could be. I'm not familiar with doom or straight but I read both your manuals and I still can't figure it out. The fact the package can load at the point of the debugger is bizarre..

**Additional details:**
- Include a link to your private config: none provided, this is reproducible in a clean doom install
- Include screenshots/casts of your issue: see trace
- If you mention key sequences, include what commands they're bound to (use `SPC
  h k KEY` or `C-h h k KEY` to inspect keys).: not relevant
- Include any warnings or errors logged to \*Messages\* (use `M-x
  view-echo-area-messages` to see it).

<details><pre>
> doom sync && doom build                     
> Synchronizing your config with Doom Emacs...
  > Regenerating envvars file at "~/doom-emacs/.local/env"
    ✓ Successfully generated "~/doom-emacs/.local/env"
  > Installing packages...
    x There was an unexpected error
      Message: Symbol's function definition is void
      Data: (void-function . transient-define-prefix)
      Backtrace:
        (transient-define-prefix blahblahblah nil "A `transient' popup for th
        (eval-buffer #<buffer  *load*-869095> nil "/Users/me/doom-emacs/.local/
        (load-with-code-conversion "/Users/me/doom-emacs/.local/straight/build/
        (load "/Users/me/doom-emacs/.local/straight/build/org-issues/org-issues
</pre></details>

**Steps to reproduce:**
Already described above. Unfortunately going to be difficult to share the package code with you.

**System information:**
<details><pre>
```
SYSTEM  type       darwin
	config     x86_64-apple-darwin18.7.0
	shell      /bin/zsh
	uname      Darwin 18.7.0 Darwin Kernel Version 18.7.0: Mon Aug 31 20:53:32 PDT 2020; root:xnu-4903.278.44~1/RELEASE_X86_64 x86_64
	path       (~/doom-emacs/bin ~/perl5/bin ~/.toolbox/bin .....elided....)
EMACS   dir        ~/doom-emacs/
	version    27.1
	build      Oct 09, 2020
	buildopts  --disable-dependency-tracking --disable-silent-rules --enable-locallisppath=/usr/local/share/emacs/site-lisp --infodir=/usr/local/Cellar/emacs-plus@27/27.1/share/info/emacs --prefix=/usr/local/Cellar/emacs-plus@27/27.1 --with-xml2 --with-gnutls --without-dbus --with-imagemagick --with-modules --with-rsvg --with-ns --disable-ns-self-contained
	features   RSVG IMAGEMAGICK GLIB NOTIFY KQUEUE ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES THREADS JSON PDUMPER LCMS2 GMP
	traits     (batch server-running envvar-file)
DOOM    dir        ~/.doom.d/
	version    2.0.9
	build      HEAD -> develop adff1aa68 2020-10-22 02:10:59 -0400
	elc-files  0
	modules    (:completion company ivy :ui doom doom-dashboard doom-quit hl-todo modeline ophints (popup +defaults) vc-gutter vi-tilde-fringe workspaces :editor file-templates fold lispy multiple-cursors snippets :emacs dired electric undo vc :checkers syntax :tools (eval +overlay) lookup lsp magit :os macos :lang emacs-lisp markdown org plantuml (rust +lsp) sh :config (default +bindings +smartparens))
	packages   (n/a)
	unpin      (n/a)
	elpa       (n/a)
```
</pre></details>

**Comment by hlissner** [comment]
When you install an Emacs package (with or without Doom), Emacs will first byte-compile its files and generate an *-autoloads.el file from `;;;###autoload` cookies from its *.el files.

This error is occurred during byte-compilation.

> (void-function . transient-define-prefix)

It indicates that the `transient-define-prefix` macro is not available when it is called. There are two ways to fix this:

1. Add `(eval-and-compile (require 'transient))` to your package, and ensure `transient` is listed as one of its dependencies in the `Package-Requires` heading, e.g.

   ```elisp
   ;; Package-Requires: ((transient "0.1.0"))
   ```

   This ensures the `transient` package is available when the package is byte-compiled.

2. Add `:no-byte-compile t` to the package's `:recipe`. This prevents the package from being compiled in the first place, _but_ you must guarantee `transient` has been loaded before this package is.


Hope that helps!

**Comment by marcbowes** [comment]
Thank you for your detailed and prompt response. I believe I've already tried both of those, at least I tried (2) verbatim.

For (1) I made sure that blahblah-pkg.el existed and correctly declared dependencies. My package was using Cask, so I did that with `cask pkg-file` and manual inspection (it does include transient).

I'm more than open to the idea I've done something wrong. What I'm at a loss for is how to figure out where it went wrong. For example, how can I inspect what doom/straight *thinks* the dependencies are?

In the meanwhile, I'll try the eval-and-compile suggestion.
